metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+            uint_kmer.kth_char_or(k_ - 1, kmer_t::char_to_uint(sequence[i]));
+            callback(sshash_to_graph_index(dict.lookup_uint(uint_kmer, false)));
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        stream_map(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        stream_map(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        stream_map(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+void DBGSSHash::map_to_nodes_with_rc(std::string_view sequence,
+                                     const std::function<void(node_index, bool)>& callback,
+                                     const std::function<bool()>& terminate) const {
+    if (terminate() || sequence.size() < k_)
+        return;
+
+    if (!num_nodes()) {
+        for (size_t i = 0; i < sequence.size() - k_ + 1 && !terminate(); ++i) {
+            callback(npos, false);
+        }
+        return;
+    }
+
+    auto stream_map = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        kmer_t uint_kmer = sshash::util::string_to_uint_kmer<kmer_t>(sequence.data(), k_ - 1);
+        uint_kmer.pad_char();
+        for (size_t i = k_ - 1; i < sequence.size() && !terminate(); ++i) {
+            uint_kmer.drop_char();
+            uint_kmer.kth_char_or(k_ - 1, kmer_t::char_to_uint(sequence[i]));
+            auto res = dict.lookup_advanced_uint(uint_kmer, true);
+            callback(sshash_to_graph_index(res.kmer_id), res.kmer_orientation);
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        stream_map(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        stream_map(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        stream_map(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+DBGSSHash::node_index DBGSSHash::traverse(node_index node, char next_char) const {
+    // TODO: if a node is in the middle of a unitig, then we only need to check the next node index
+    return kmer_to_node(get_node_sequence(node).substr(1) + next_char);
+}
+
+DBGSSHash::node_index DBGSSHash::traverse_back(node_index node, char prev_char) const {
+    // TODO: if a node is in the middle of a unitig, then we only need to check the previous node index
+    std::string string_kmer = std::string(1, prev_char) + get_node_sequence(node);
+    string_kmer.pop_back();
+    return kmer_to_node(string_kmer);
+}
+
+void DBGSSHash::adjacent_outgoing_nodes(node_index node,
+                                        const std::function<void(node_index)>& callback) const {
+    assert(node > 0 && node <= num_nodes());
```

Comment:
Is there a reason to have this check here specifically, but not in other places where `node_index node` is an argument?

Reply:
no, this check should actually be on all such functions